File: example-clients/node/sse_client.mjs
#!/usr/bin/env node
// Consume a WorldSignal subscription as Server-Sent Events.
//
// The browser EventSource API can't set an Authorization header, so this reads
// the stream with fetch() and parses `data:` lines itself — letting it use the
// header (preferred) instead of the ?api_key= query param. Dependency-free.
// Env: WS_API_BASE, WS_API_KEY, WS_SUBSCRIPTION, WS_SINCE, WS_MAX.
const BASE = process.env.WS_API_BASE || "http://localhost:4000";
const KEY = process.env.WS_API_KEY || "";
const SUB = process.env.WS_SUBSCRIPTION || "demo-stream";
const SINCE = process.env.WS_SINCE || "0";
const MAX = Number(process.env.WS_MAX || "0");

async function main() {
  if (!KEY) { console.error("WS_API_KEY is required"); process.exit(2); }
  const url = `${BASE}/v1/stream/sse?subscription=${SUB}&since=${SINCE}`;
  console.error(`[sse] connecting to ${url}`);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${KEY}` } });
  const reader = res.body.getReader();
  const dec = new TextDecoder();
  let buf = "";
  let seen = 0;
  for (;;) {
    const { value, done } = await reader.read();
    if (done) break;
    buf += dec.decode(value, { stream: true });
    let nl;
    while ((nl = buf.indexOf("\n")) >= 0) {
      const line = buf.slice(0, nl).replace(/\r$/, "");
      buf = buf.slice(nl + 1);
      if (!line.startsWith("data:")) continue; // skip id:/event:/comments
      const ev = JSON.parse(line.slice(5).trim());
      const d = ev.data || {};
      console.log(`[sse] ${(d.severity || "?").padEnd(8)} ${d.country || "--"}  ${d.title || ""}`);
      if (MAX && ++seen >= MAX) { console.error(`[sse] received ${seen} event(s), exiting`); return; }
    }
  }
}
main().catch((e) => { console.error(e); process.exit(1); });

File: example-clients/ruby/sse_client.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Consume a WorldSignal subscription as Server-Sent Events — stdlib only.
# Reads the stream incrementally and parses `data:` lines, using the
# Authorization header (EventSource can't set one). Env: WS_API_BASE,
# WS_API_KEY, WS_SUBSCRIPTION, WS_SINCE, WS_MAX.
require 'json'
require 'net/http'
require 'uri'

BASE = ENV.fetch('WS_API_BASE', 'http://localhost:4800')
KEY = ENV.fetch('WS_API_KEY', '')
SUB = ENV.fetch('WS_SUBSCRIPTION', 'demo-stream')
SINCE = ENV.fetch('WS_SINCE', '0')
MAX = ENV.fetch('WS_MAX', '0').to_i

abort('WS_API_KEY is required') if KEY.empty?

uri = URI("#{BASE}/v1/stream/sse?subscription=#{SUB}&since=#{SINCE}")
warn("[sse] connecting to #{uri}")
req = Net::HTTP::Get.new(uri)
req['Authorization'] = "Bearer #{KEY}"

seen = 0
buf = +''
Net::HTTP.start(uri.hostname, uri.port) do |http|
  http.request(req) do |res|
    res.read_body do |chunk|
      buf << chunk
      while (nl = buf.index("\n"))
        line = buf.slice!(0..nl).chomp
        next unless line.start_with?('data:')

        ev = JSON.parse(line[5..].strip)
        d = ev['data'] || {}
        puts format('[sse] %-8s %s  %s', d['severity'] || '?', d['country'] || '--', d['title'] || '')
        seen += 1
        if MAX.positive? && seen >= MAX
          warn("[sse] received #{seen} event(s), exiting")
          exit(0)
        end
      end
    end
  end
end
